verifier/src/certs.rs

```diff
@@ -56,20 +59,50 @@ pub struct VerifiedCertificate<'a> {
 }
 
 impl<'a> UnverifiedCertificate<'a> {
-    /// Verify the certificate signature.
-    pub fn verify(self, key: &VerifyingKey) -> Result<VerifiedCertificate<'a>> {
-        let tbs_length = self.certificate.tbs_certificate.encoded_len()?;
-        let tbs_size = u32::from(tbs_length) as usize;
-        let tbs_contents = &self.der_bytes[TBS_OFFSET..tbs_size + TBS_OFFSET];
-        key.verify(tbs_contents, &self.signature)
-            .map_err(|_| Error::SignatureVerification)?;
+    /// Verify the certificate signature and time are valid.
+    ///
+    /// # Arguments
+    /// - `key` - The public key to verify the certificate signature with
+    /// - `unix_time` - The duration since UNIX_EPOCH. This value can have
+    ///   `nanos`.
+    pub fn verify(
+        self,
+        key: &VerifyingKey,
+        unix_time: Duration,
+    ) -> Result<VerifiedCertificate<'a>> {
+        self.verify_time(unix_time)?;
+        self.verify_signature(key)?;
+
         Ok(VerifiedCertificate {
             _der_bytes: self.der_bytes,
             _certificate: self.certificate,
             _signature: self.signature,
             _key: self.key,
         })
     }
+
+    fn verify_signature(&self, key: &VerifyingKey) -> Result<()> {
+        let tbs_length = self.certificate.tbs_certificate.encoded_len()?;
+        let tbs_size = u32::from(tbs_length) as usize;
+        let tbs_contents = &self.der_bytes[TBS_OFFSET..tbs_size + TBS_OFFSET];
+        key.verify(tbs_contents, &self.signature)
+            .map_err(|_| Error::SignatureVerification)?;
+        Ok(())
+    }
+
+    fn verify_time(&self, unix_time: Duration) -> Result<()> {
+        let validity = &self.certificate.tbs_certificate.validity;
+        let not_before = validity.not_before.to_unix_duration();
+        let not_after = validity.not_after.to_unix_duration();
+
+        // Per https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.5 time is
+        // inclusive
+        if (not_before..=not_after).contains(&unix_time) {
+            Ok(())
+        } else {
+            Err(Error::CertificateExpired)
+        }
+    }
```

Comment:
Could we do something like:
```suggestion
        if not_before <= unix_time && unix_time <= not_after {
            Ok(())
        } else {
            Err(Error::CertificateExpired)
        }
    }
```
Creating a `Range` object to perform an equivalency check seems a little wonky to me.

Reply:
Changed logic to have a clearer error for before time.